Tap14 - Round 2, seven years in the making

isaacs opened the pull request:
The goal this time around will be to make small uncontroversial changes, which can step us into paving the paths laid by node-tap, Test::More, tappy, and many others over the years.  There will be no innovation here, just cataloging what exists.

Goals:

* No changes will be accepted UNLESS they are already widely adopted by multiple implementations representing a majority of the TAP userbase.

* No changes will be blocked IF they are already widely adopted by multiple implementations representing a majority of the TAP userbase.

* Specification should remove as much ambiguity as possible, while maintaining backwards compatibility with existing TAP providers and consumers.

Features/Changes Anticipated:

* Change the version from 13 to 14 (of course)
* Require that yaml diagnostic blocks be actual yaml, stipulate that they are to be indented 2 spaces.
* Specify child tests (both indented and {}-wrapped buffered), indented 4 spaces.
* Specify pragmas (at least, define the syntax, even if the specific keys are left up to implementations)
* Clean up and clarify ambiguous language.

isaacs committed:
start the v14 spec with the v13 spec

The goal this time around will be to make small uncontroversial
changes, which can step us into paving the paths laid by node-tap,
Test::More, tappy, and many others over the years.  There will be no
innovation here, just cataloging what exists.

Goals:

* No changes will be accepted UNLESS they are already widely adopted by
  multiple implementations representing a majority of the TAP userbase.

* No changes will be blocked IF they are already widely adopted by
  multiple implementations representing a majority of the TAP userbase.

* Specification should remove as much ambiguity as possible, while
  maintaining backwards compatibility with existing TAP providers and
  consumers.

Features/Changes Anticipated:

* Change the version from 13 to 14 (of course)
* Require that yaml diagnostic blocks be actual yaml, stipulate that
  they are to be indented 2 spaces.
* Specify child tests (both indented and {}-wrapped buffered), indented
  4 spaces.
* Specify pragmas (at least, define the syntax, even if the specific
  keys are left up to implementations)
* Clean up and clarify ambiguous language.

isaacs committed:
TAP14: update version number

Hopefully this much we can agree upon, at least ;)

Note that a parser is still technically allowed to interpret a `TAP
version 13` stream as TAP14 and be compliant with this spec, but this
behavior is not required.  (In other words, they can also throw that out
as an error, and also be compliant.)  Since the goal is to codify
existing behavior and add little to it, and all current "TAP13"
implementations allow `TAP version 13`, this follows logically.

isaacs committed:
TAP14: put language markers on code fence blocks

isaacs committed:
TAP14: First cleanup pass

- Provide rough EBNF grammar for the tap document.
- Allow Harnesses to accept `TAP version 13` and still be TAP14
  compliant.
- Group all Test Point definitions together.
- Include RFC2119 language note.
- Clarify Harness behavior without being implicitly tied to perl command
  line test runners.
- Clarification of TODO and SKIP directives, specify that they MUST NOT
  be treated as a test failure, even if the Test Point is `not ok`.
- Specify that YAML blocks must be indented by 2 spaces.
- Provide guidance around test exit status, handling of invalid TAP
  lines, and other Harness behavior.

isaacs committed:
TAP14: Add (temporary) status/goals section up top

isaacs committed:
TAP14: unknown directives are a thing

isaacs commented:
At this point, basically every current TAP13 harness should also be a valid TAP14 harness according to this spec, and with the exception of the Version line, every TAP14 document is also a valid TAP13 document (though not vice versa, owing to additional strictness in the YAML indentation).

If we can accept this much and move forward, the remaining work is just to specify how child tests are defined.  Which is the biggest (really, _only_) feature we're talking about adding for TAP14, and since there's already fairly broad consensus on what we support, I hope it should be straightforward.

isaacs commented:
@exodist @Leont @mblayman ptal

Leont commented on tap-version-14-specification.md:
````diff
@@ -0,0 +1,659 @@
+```ebnf
+TAPDocument := Version Plan Body | Version Body Plan
+Version     := "TAP version 14\n"
+Plan        := (Number) ".." (Number) (" # " Reason)? "\n"
````
Is the first number ever not 1?

isaacs replied:
There are examples I've come across in the wild of people trying to use http://npm.im/tap-parser on stuff like:

```tap
TAP version 13
4..6,10..12
ok 4
ok 6
ok 5
ok 10
ok 12
ok 11
```

I compromised and said, ok, _one_ range which can start and end wherever, and node-tap itself will only ever _emit_ a plan starting with 1.  The tap 13 specification doesn't say it _has_ to start with 1, but at the time I pushed back on allowing multiple ranges.  I don't know what they were using to emit their test data, I think just a custom thing.

If we wanna say that TAP14 requires plans be `1..N`, then I'm definitely on board with that!

isaacs replied:
Definitely `6..4` shouldn't be allowed, so `1..0` is the only case where the "end" is lower than the "start".

isaacs replied:
tap4j (@kinow) and node-tap (@isaacs) say plans _must_ start with 1.  Can I get a :+1: from perl (@exodist) or python (@mblayman) on this?  (Or if you think we should only go as far as "producers _should_ start plans with 1"/"harnesses _may_ reject/ignore plans if they don't start with 1", I'm fine with that, as well.)

isaacs replied:
2fd8aaf895e58210aa173faf0826ba40bbc6cb4d

Leont replied:
Both Perl harnesses require it to start with a 1

Leont commented on tap-version-14-specification.md:
````diff
@@ -0,0 +1,659 @@
+
+TAP14's general grammar is:
+
+```ebnf
````
Can EBNF accurately describe how (nested) subtests work? My intuition says not, but I'd hope to be proven wrong.

isaacs replied:
Probably somehow, but in my experience, using EBNF to describe something like that is "wrong tool for the job".  It's more for humans than for machines, so you can just sorta say "all the same stuff that's in a TAP document, but indented 4 spaces" and be done with it.

isaacs replied:
89380697f76e78baddb9d33d73e8c937a9bcd9e3

Leont commented:
> Goals:
>
>  * No changes will be accepted UNLESS they are already widely adopted by multiple implementations representing a majority of the TAP userbase.
>
> * No changes will be blocked IF they are already widely adopted by multiple implementations representing a majority of the TAP userbase.
>
> * Specification should remove as much ambiguity as possible, while maintaining backwards compatibility with existing TAP providers and consumers.

These goals sound sensible to me.

mblayman commented on tap-version-14-specification.md:
```diff
@@ -0,0 +1,659 @@
+comment lines.  That is, a Harness _must_ ignore any unrecognized comment
+lines.
+
+### Pragmas
```
I think this one is beyond the scope that was discussed, but I don't have a problem with it. It sounds useful.

mblayman replied:
I like that idea. Since pragmas seem to be control mechanisms for harnesses that want to change the behavior/interpretation of a stream, changing the language to use _may_ provides an out for harnesses that want to ignore pragma lines entirely.

I think this would be especially true since there is no well defined set of pragmas that a harness is expected to enforce.

Leont replied:
I think it would be useful to be able to make a pragma as "if you don't understand this you should treat it as an error" (exclamation mark feels like the right marker for this). But that may be for the next version.

isaacs replied:
943e7d733cbe9e4b0bf337d3cfdb35058f32242e

isaacs replied:
943e7d733cbe9e4b0bf337d3cfdb35058f32242e

mblayman commented on tap-version-14-specification.md:
```diff
@@ -0,0 +1,659 @@
+
+The original TAP documentation (of which this is a hacked about version)
+was written by Andy Lester, based on the original Test::Harness
+documentation by Michael Schwern.  This version was written by Andy
```
@isaacs, I think you could attach your name to this version.

isaacs replied:
Haha, yeah, didn't get that far down yet, but I'll definitely update it with all who contribute to this version (myself included).

isaacs replied:
fc4b81311e9b55baefda6be55ec2539f992695c4

mblayman reviewed (approved):
I thought this version would include subtests, but perhaps we're holding back on that until there is consensus on https://github.com/TestAnything/Specification/issues/2. Even in the absence of that, I think this is a step in a positive direction.

kinow reviewed (commented):
Great work @isaacs , thanks a lot for writing this! +1 to adding your name under authors as suggested in another comment.

I added a lot of comments, but as a non native speaker, some may be irrelevant. Other comments were about differences between TAP13 and TAP14, or about the grammar.

But this version looks really promising! Once the comments are resolved, I'll hit approve and prepare to update the Java implementation `tap4j` and the Jenkins plug-in to support it too. I think TAP14 wouldn't break anything in `tap4j`, but pragmas are ignored as invalid line, so I'll just change it to parse them correctly.

Great work again, bravo :clap: 
-Bruno

isaacs commented on tap-version-14-specification.md:
```diff
@@ -0,0 +1,659 @@
+these programs, and handles it in some way.
+
+A harness that is collecting output from test programs _should_ read and
+interpret TAP from the process's standard output, not standard error.
```
Probably easiest to rewrite the sentence this way to reduce ambiguity:

```suggestion
A harness that is collecting output from a test program _should_ read and
interpret TAP from the process's standard output, not standard error.
```

isaacs replied:
7b6b2093d57a9a6c15692229a5d191cca678d48c

isaacs commented on tap-version-14-specification.md:
```diff
@@ -0,0 +1,659 @@
+At the present time (March 2022), most popular TAP generating utilities
+always start Plan lines at 1.
+
+### Test Points
```
The terms "test point", "test line", and "test result" are used somewhat interchangeably in TAP13.  I just picked one arbitrarily, any of them work.  I recall seeing it called "test point" in other perl testing literature, which I'd copied into the JS implementation.

kinow replied:
Got it. I remember test line, and may have read test point but my memory is failing me. Thanks!

isaacs commented on tap-version-14-specification.md:
````diff
@@ -0,0 +1,659 @@
+    ok
+    ```
+
+    has five tests. The sixth is missing. Test::Harness will generate
````
Ah, yeah I see that's unclear, good catch.

I think the answer is that this tap:

```tap
TAP version 13
1..999
ok
ok
<eof>
```

_is_ valid tap, but _is not_ a successful test run (plan not satisfied).  So the harness should treat it as a failure, but should not throw the whole thing out as invalid, necessarily.

kinow replied:
Ah, yes, the plan not satisfied here makes sense :+1: ! Good explanation, maybe we can just paste that replacing the text in the comment. Thanks.

isaacs replied:
6d619d29998c8a71604a494e31a202ef14ad20c3

isaacs commented on tap-version-14-specification.md:
```diff
@@ -0,0 +1,659 @@
+    Descriptions _should_ be separated from the Test Point Status and Test
+    Point ID by the string `" - "`, in order to prevent confusing a numeric
+    description with a Test Point ID.  However, the `" - "` separator is
+    _optional_.
```
I'll clarify.  The producer _should_ separate the description with a `- `, but a harness _must_ treat `ok 1 - foo` and `ok 1 foo` identically.

kinow replied:
>but a harness must treat ok 1 - foo and ok 1 foo identically.

:bow: ah, sounds good to me (both your explanation, and also the action of treating both identically.)

isaacs replied:
f1803dd8f1acfa088cdbc1a109f73b91f30d2a98

isaacs commented on tap-version-14-specification.md:
```diff
@@ -0,0 +1,659 @@
+
+##### SKIP tests
+
+If the directive starts with `# SKIP`, the test is counted as a todo test,
```
Copypasta error, good catch.

isaacs replied:
f843c1ad656a8582bb8df4e64936d3a10ebbbecc

isaacs commented on tap-version-14-specification.md:
```diff
@@ -0,0 +1,659 @@
+TAPDocument := Version Plan Body | Version Body Plan
+Version     := "TAP version 14\n"
+Plan        := (Number) ".." (Number) (" # " Reason)? "\n"
+Body        := (TestPoint | BailOut | Pragma | Comment | Anything | Empty)*
```
I think technically (as the draft stands right now) in `ok 1 - test # comment`, the `# comment` is an unrecognized directive.  I don't love that.

But that brings up an interesting point to investigate, how to have `#` in a test point name?  This gets weird:

```tap
$ node -e 'require("tap").pass("ok # commenty")'
TAP version 13
ok 1 - ok # commenty
1..1
# time=1.489ms
```

Node's `tap-parser` will interpret any unrecognized directive as part of the description (like you suggest)

```
$ node -e 'require("tap").pass("ok # commenty")' | npx tap-parser
[
  [ 'version', 13 ],
  [
    'assert',
    Result { ok: true, id: 1, name: 'ok # commenty', fullname: '' }
  ],
  ...
```

But then making that test TODO breaks things a bit...

```
$ node -e 'require("tap").pass("ok # commenty", { todo: true })'
TAP version 13
ok 1 - ok # commenty # TODO
1..1
# time=1.48ms
```

```
$ node -e 'require("tap").pass("ok # commenty", { todo: true })' | npx tap-parser
[
  [ 'version', 13 ],
  [
    'assert',
    Result {
      ok: true,
      id: 1,
      name: 'ok # commenty # TODO',
      fullname: ''
    }
```

So, it seems like some options are:

- Specify a way to escape `#` chars, so they're not parsed as directives, and say that producers _should_ escape any `#` characters in test descriptions.  (This seems like a good idea regardless, but I'm not sure if there's consensus on it.  Node-tap will not parse directives that use `\` to escape the `#`.)
- Allow multiple directives (then how can you have # in the `TODO` description?) I don't really want to do this, it gets weird fast 😓
- Do not allow `#` in the test point description (this seems especially ruthless 😬)
- Specify that everything after the first `#` is _the_ directive, and tell harnesses to include unrecognized directives in the description (then this breaks `ok 1 - name # comment # TODO`)
- Specify that everything after the _last_ `#` is _the_ directive (then this breaks `ok 1 - name # TODO # reason`)
- Specify that harnesses should search the string for the first known directive, and treat it as the only directive for that test point, and everything else up to that point is part of the description.  (So `ok 1 - desc # todo # skip` is a todo test with the reason set to `"# skip"`, and `ok 1 - desc # skip # todo` is a skip test with the reason set as `# todo`).
- Just document the edge case and leave it up to harness/producer implementations 🤷  (I'm fine with this, it just kicks the can down the road for implementations to come to consensus for TAP15.)

kinow replied:
I remember seeing test points with comments somewhere some long time ago, probably in some Perl examples/projects. So I implemented that in tap4j, but it was part of the issues I had when trying to write a formal grammar (ended up writing a horrendous regular expression parser).

I think I'd lean towards either treating it as an ignored directive attached to a test point, or as a comment attached to a test point. It would be easier to avoid hashes within hashes, but in `tap4j` that could happen as that was used in Java, where you can get a method like in a comment like `not ok 1 # SKIP see comment in org.example.Class#methodName(int, String)` (text like this definitely appears in the YAMLish, but users could add it to directives/comments too I guess.)

mblayman replied:
In my reading of the TAP13 spec, the language used was "*starts with* # TODO". In tappy, I took this to mean:

* The first `#` is the start of a directive.
* There can be only one recognized directive per test point.
* Only TODO and SKIP are recognized (with the appropriate case sensitivity and permission of characters after "SKIP" like "Skipped" applied)

The language in the spec calls the stuff after as "explanation" or "reason" depending on the context, so I normalized that part as a "description."

Given all that, I think I would lean toward using the second to last bullet to recognize only a single directive and treat everything after it as the description. Things that don't look like a TODO or SKIP would be an unknown directive (or ignored outright).

Leont replied:
> Specify a way to escape # chars, so they're not parsed as directives, and say that producers should escape any # characters in test descriptions. (This seems like a good idea regardless, but I'm not sure if there's consensus on it. Node-tap will not parse directives that use \ to escape the #.)

Test::Harness allows for `#`(or `\`) to be escaped with `\`, and has done so since at least 2007

isaacs replied:
Promoting this to an issue: https://github.com/TestAnything/Specification/issues/26

isaacs commented on tap-version-14-specification.md:
```diff
@@ -0,0 +1,659 @@
+
+Harnesses _must not_ treat unrecognized Pragma keys as a test failure.
+However, they _may_ warn if a Pragma is unrecognized, or fail if the
+named flag cannot be set for some reason.
```
Node-tap treats `pragma -bail # comment` as a non-TAP line.

kinow replied:
I wonder what others think. I think I could probably document pragmas with comments above it, but I guess some users would/could also try to put it as a comment/directive next to it? If the spec doesn't say it's invalid and must be treated as non-TAP line, I'd probably add support to it in `tap4j`.

isaacs replied:
If tap4j doesn't already support pragmas at all today, I'd recommend not adding trailing # comments.  Trying to be lazy and limit how much _new_ work we make for ourselves here 😅  We still have child tests to specify.

As written so far in the draft, `pragma -bail # comment` _would_ be under "anything else", because `"bail # comment"` does not match `/[a-zA-Z0-9_-]+/`.

isaacs replied:
d5a477e72d449247fd46062bbead773949c976e2

isaacs commented on tap-version-14-specification.md:
```diff
@@ -0,0 +1,659 @@
+- Directive (only when necessary)
+
+
+#### DIRECTIVES
```
My OCD and your OCD are in alignment here.  Just copying from the TAP13 spec.  I'll make it all consistent before it's done 😅

Leont replied:
That's a convention dating back to manpages, it does feel a bit dated nowadays.

isaacs replied:
Back when we walked uphill in the snow both ways to school and our only stylesheets were troff escape codes.

isaacs replied:
ba460ac138568f2103cd09b1780acb98d15a82d7

isaacs commented on tap-version-14-specification.md:
```diff
@@ -0,0 +1,659 @@
+a database, it doesn't know exactly how many servers it will need to ping.
+Thus, the test count is declared at the bottom after all the test points
+have run. Also, two of the tests fail. The YAML block following each
+failure gives additional information about the failure that may be
```
I only highlight the `may`s that are used in the technical sense of RFC 2119.

isaacs commented on tap-version-14-specification.md:
````diff
@@ -0,0 +1,659 @@
+1..0 # skip because English-to-French translator isn't installed
+```
+
+## Got spare tuits?
````
I'm glad you're taking the time to give this such a careful read, it's very helpful having a non-native english speaker to find all the things that don't seem strange to me, since I've grown up with this strange language.

kinow replied:
>In english, sometimes you say that you'll do something "when you get around to it". That can be interpreted as "when you get a round tuit". No one knows what a "tuit" is, but if you find one that's round and doesn't have any corners, now you can do all those things you put off.

Aaaahhh!!! Thanks a lot for the explanation! And that's a good joke!

mblayman replied:
Ha! I'm a native English speaker and I had no idea what tuit was. 😄  I assumed it was some Perl joke from years ago and rolled with it.

Leont replied:
> Ha! I'm a native English speaker and I had no idea what tuit was. smile I assumed it was some Perl joke from years ago and rolled with it.

I don't think Perl was the origin of it, but actually having round tuits made (wooden coins with tuit on it) was definitely a Perl joke from years ago 😋

isaacs replied:
@Leont That joke predates Perl by at least 5 decades, maybe more.  My grandfather had a round ceramic disc with "tuit" written on it from some time around the great depression.  The height of humor in early 20th century America, apparently.

isaacs replied:
ba460ac138568f2103cd09b1780acb98d15a82d7

Leont replied:
Are you calling us old-fashioned now? 😅

kinow commented on tap-version-14-specification.md:
````diff
@@ -0,0 +1,659 @@
+    ```
+
+    Test Points _may_ be output in any order, but any Test Point ID
+    provided _must_ be within the range described by the Plan.
````
Ah, _may_, and not _must_. Sorry, didn't read it properly. Marking as resolved.

isaacs commented:
Thanks for all the feedback, I'll get some time to take another pass through this either today or this weekend.

isaacs committed:
TAP14: improve pseudo-EBNF

isaacs committed:
TAP14: plans must start with 1..

isaacs committed:
TAP14: loosen/clarify expectations re Pragma lines

isaacs committed:
TAP14: Copyright, authorship, and licensing

isaacs committed:
TAP14: language cleanup on stderr harness behavior

isaacs committed:
TAP14: it's -> its

isaacs commented on tap-version-14-specification.md:
```diff
@@ -0,0 +1,659 @@
+The Plan lists the range of test point IDs that are expected in the TAP
+stream.  It can also optionally contain a comment prefixed by a `#`.
+
+It's basic grammar is:
```
7595f7a35b31f5e2ea18f652dbafdfbc2f863cd4

isaacs committed:
TAP14: specify what counts as 'blank' in a blank line

isaacs committed:
TAP14: clarification of test point IDs, plan ranges, and failures

isaacs committed:
TAP14: clarify optionality of ' - ' in Test Point Description

isaacs committed:
TAP14: <code> the #

isaacs commented on tap-version-14-specification.md:
```diff
@@ -0,0 +1,659 @@
+
+#### DIRECTIVES
+
+Directives are special notes that follow a # on the Test Point line. Only
```
2ffc0e4d55667b9514b137339b185bb5141d8cf6

isaacs committed:
TAP14: clarify case-sensitivity of directives

isaacs commented on tap-version-14-specification.md:
```diff
@@ -0,0 +1,659 @@
+#### DIRECTIVES
+
+Directives are special notes that follow a # on the Test Point line. Only
+two are currently defined: `TODO` and `SKIP`. These two keywords are not
```
fff58d6aeecd0ad25ac433f6625c28610b0e9b7c

isaacs committed:
TAP14: copypasta error s/todo/skip/

isaacs committed:
TAP14: Consistent capitalization, refresh ancient joke

isaacs committed:
TAP14: make pragmas kinda strict

isaacs committed:
TAP14: missed a manpage-esque caps fix

isaacs committed:
TAP14: tap is utf8

Fix: https://github.com/TestAnything/Specification/issues/12

isaacs committed:
TAP14: MORE CAPITALIZATION fixes

isaacs committed:
TAP14: specific instructions for escaping # and \

Re discussion in https://github.com/TestAnything/Specification/issues/26

isaacs commented:
Ok, at this point all the feedback has been incorporated and resolved according to how we currently do things.  If you commented on the previous draft, PTAL.

I _believe_ the only thing left to do is nested subtests, which is the biggest change, but now we're starting from a more clearly defined base.

I'll let it sit for a week or two to give some more room for comments to come in, and also because that's probably when I'll get time to work on it again :)

isaacs committed:
TAP14: <code> Test::Harness

isaacs committed:
TAP14: thank Jonathan Kingston

isaacs committed:
TAP14: formatting of Test Point sections

isaacs committed:
TAP14: (provisional) no repeating Test Point IDs

https://github.com/TestAnything/Specification/issues/27

mblayman commented on tap-version-14-specification.md:
```diff
@@ -0,0 +1,807 @@
+testing modules a test harness. TAP started life as part of the test
+harness for Perl but now has implementations in C/C++, Python, PHP, Perl,
+JavaScript, and probably others by the time you read this.  This document
+describes version 14 of TAP. Go to TAP to read about previous versions.
```
```suggestion
describes version 14 of TAP.
```
nit: I think that last sentence is a hangover from when the spec used to be on a wiki.

mblayman commented on tap-version-14-specification.md:
```diff
@@ -0,0 +1,807 @@
+
+##### `SKIP` tests
+
+If the directive starts with `# SKIP`, the test is counted as a skipped
```
To be backwards compatible with TAP13 and support `# Skipped`, I think we need the directive to support the regex that permits non-whitespace characters immediately next to SKIP, which is defined in that version of the spec.

> If the directive starts with # SKIP, the test is counted as having been skipped. If the whole test file succeeds, the count of skipped tests is included in the generated output. The harness should report the text after # SKIP\S*\s+ as a reason for skipping.

isaacs replied:
Ahhhh yes. Node-tap doesn't support this today, but I see it should, according to tap13 spec. This is why clear specs are useful! I'll update and fix node-tap.

mblayman reviewed (commented):
@isaacs, I appreciate all your hard work on this. It's looking good and I'm excited to see subtests defined so we can finish off this version. I commented on a couple of minor things that I found.

isaacs committed:
TAP14: remove wiki holdover language

Co-authored-by: Matt Layman <matthewlayman@gmail.com>

isaacs committed:
TAP14: SKIP and TODO can have additional non-whitespace characters

mblayman commented on tap-version-14-specification.md:
```diff
@@ -0,0 +1,837 @@
+##### Backwards Compatibility Note
+
+For backwards compatibility with earlier incarnations of TAP, Harnesses
+_must_ accept additional non-whitespace characters following the the
```
```suggestion
_must_ accept additional non-whitespace characters following the
```

isaacs committed:
TAP14: require directive #'s to be preceded by whitespace

This prevents eg 'https://example.com/x.html#skip' being treated as a
directive, since #'s occur in urls and other contexts.

Of course, a producer _should_ escape the # in that case, but that can't
be relied upon, so the Harness _should_ handle it.

isaacs committed:
TAP14: fix fix the the extra extra word word

Co-authored-by: Matt Layman <matthewlayman@gmail.com>

isaacs commented:
TODO:

- [ ] clean up directive specification in light of https://github.com/TestAnything/Specification/issues/28
- [ ] child tests 😬 (that's the big one)

isaacs committed:
TAP14: clarify space around # in directives

Fix: https://github.com/TestAnything/Specification/issues/28

kinow reviewed (commented):
Looking good!!! :champagne: :trophy: 

I think it's only missing subtests now, then it should be ready for a final review round. Thanks @isaacs !

martinheidegger commented on tap-version-14-specification.md:
````diff
@@ -0,0 +1,903 @@
+and any text after `TODO\S*\s+` is the explanation.
+
+```tap
+not ok 14 # TODO bend space and time
````
Should the implementations note `# TODO` tests as `not ok`?

martinheidegger commented on tap-version-14-specification.md:
````diff
@@ -0,0 +1,903 @@
+For example:
+
+```tap
+TAP version 13
````
`version 14`?

isaacs replied:
Ah, yeah, that's confusing.  The only "version 13" should be in the bit where we say that version 13 is also fine.

isaacs committed:
TAP14: use version 14 except where relevant

isaacs committed:
TAP14: Subtests

Fix: https://github.com/TestAnything/Specification/issues/29

isaacs commented:
First draft of subtests are in.  PTAL when you get the chance.  Thanks!

kinow commented on tap-version-14-specification.md:
````diff
@@ -0,0 +1,1167 @@
+    1..2
+    ```
+
+Subtests are designed with graceful fallback for TAP 13 harnesses in mind.
````
I think we are using `TAP14` for referring to the TAP 14 specification. In some parts of this doc we have `TAP13`, but also `TAP 13`. I think we need one last pass over the doc to use a single format, as with TAP14.

kinow commented on tap-version-14-specification.md:
```diff
@@ -0,0 +1,1167 @@
+### Bare Subtests
+
+In its simplest form, a Subtest is introduced by a 4-space indented Test
+Point, Version, Plan or Pragma line.
```
>or Pragma line.

Maybe add `(when the harness supports it)`, or something similar? I think it is implicit as before we say that pragmas are optional for implementers, but :man_shrugging:

kinow commented on tap-version-14-specification.md:
```diff
@@ -0,0 +1,1167 @@
+Harnesses _must not_ allow Pragmas set in Subtests to affect the behavior
+of the parser with respect to the parent TAP Document.
+
+For example, given a Harnesses where a `strict` Pragma will cause it to
```
s/a Harnesses/a Harness

kinow commented on tap-version-14-specification.md:
```diff
@@ -0,0 +1,1167 @@
+    1. Producers _must_ communicate the intended status of the subtest
+       (pass/fail/todo/skip/etc.) by assigning these semantics to the
+       correlated Test Point.
+    2. Harneses _should_ treat the entire subtest as either a pass or fail
```
s/Harneses/Harnesses

kinow commented on tap-version-14-specification.md:
```diff
@@ -0,0 +1,1167 @@
+<andy@petdance.com>, Andy Armstrong <andy@hexten.net>.
+
+This specification is released under the [Artistic License
+2.0](https://opensource.org/licenses/Artistic-2.0)
```
TIL about Artistic-2.0. I [think](https://www.acc.com/sites/default/files/2021-01/Crowell%20%26%20Moring%20-%20ACC%20SFBA%20Free%20and%20Open%20Source%20License%20Summary%20-%20Features%20that%20Define%20Strategy.pdf) it's similar to creative commons, but people using the content must include license, disclose source, and tell us about any changes made. :+1:

isaacs replied:
Yes, essentially it's permissive but you can't change it (meaningfully) and claim it's the same thing.  Have to either send the changes upstream, or call it by a new name.

kinow reviewed (approved):
Left a few comments @isaacs , but no blockers. Great job! Thank you.

isaacs committed:
TAP14: corrections from @kinow

isaacs committed:
TAP14: pragmas only introduce subtests if they're supported

mblayman commented on tap-version-14-specification.md:
```diff
@@ -0,0 +1,1168 @@
+   until a matching test point (same name, or no name in either) at parent
+   level is found, treating any other un-indented lines as non-TAP, and
+   fail the test if a matching test point is not found.
+9. If subtest comment is not provided, Harnesses _must_ treat the next Test
```
```suggestion
9. If Subtest Comment is not provided, Harnesses _must_ treat the next Test
```
nit: adding a bit of internal consistency to this numbered list.

isaacs replied:
Yes, "Subtest Comment" is a Defined Term, so it should be capitalized whenever it's used in this manner. A specification is the place to indulge all your worse nit-picking impulses :+1:

mblayman reviewed (approved):
The definition of subtests that you added seems to match up with the discussion. I think it provides enough specificity to meaningfully describe subtests.

As someone with an implementation that *doesn't* handle subtests, I appreciate the detail and examples provided here.

👍

isaacs committed:
TAP14: "Subtest Comment" is a Defined Term

Co-authored-by: Matt Layman <matthewlayman@gmail.com>

perlpunk reviewed (commented):
Maybe it would be good to also specify the YAML version. That would be 1.2.

isaacs committed:
TAP14: specify YAML v1.2

Re: https://github.com/TestAnything/Specification/pull/25#pullrequestreview-922477041

isaacs commented:
Haven't had any other input, and as far as I can tell, this is just about complete.  Unless there's any objection, let's call it done 2 weeks from now? (ie, Thursday, April 14.)

mblayman commented:
I like that plan. That's a reasonable amount of time to work through any final comments that might pop up.

isaacs commented:
Alright, the deadline came and no one raised any new issues.  I'll clean up the commit history and we can make it official.

isaacs committed:
TAP14: no longer a draft!

isaacs commented:
Can I get a :+1: on https://github.com/TestAnything/testanything.github.io/pull/164 to add to the website?

eli-schwartz commented on tap-version-14-specification.md:
````diff
@@ -0,0 +1,1161 @@
+TAP version 13
+```
+
+Harnesses _may_ treat any TAP stream lacking a version as a failed test.
````
I implemented this in Meson as a passing test which verbosely warns you not to do this. I got a response:

https://github.com/mesonbuild/meson/pull/11186#issuecomment-1362809587

> Meson is able to parse TAP12 and TAP13, therefore it does not have to warn if the version line is missing.

> IMO the "Harnesses may treat any TAP stream lacking a version as a failed test" is just silly. The point of TAP is exactly that it is easy to produce without relying on a library that handles the versioning and formatting for you, so good luck enforcing that with dozens of TAP producers.

Thoughts?

eli-schwartz replied:
More importantly:

> In fact the test-anything.org page says:
> 
> > Here’s what a TAP test stream looks like:
> > ```
> > 1..4
> > ok 1 - Input file opened
> > not ok 2 - First line of the input valid
> > ok 3 - Read the rest of the file
> > not ok 4 - Summarized correctly # TODO Not written yet
> > ```

Which regardless of anything else seems to be a problem -- the homepage of https://testanything.org/ describes a TAP stream without the soft-mandatory version, so it is valid TAP 12 but a TAP 14 harness can report it as a test failure. Maybe this should include a version...

bonzini replied:
Thanks for bringing it up here @eli-schwartz. In my opinion the correct reading of the spec is:

- a TAP harness MUST treat a missing version line as version 12

- a TAP harness MAY reject versions below any limit they would like to apply (even though it's not a good idea :))

The latter does imply rejecting input without a version line (which puts the limit at version 13), but does not mean that harnesses should warn if the version line is absent.

isaacs replied:
Let's discuss this in a new issue, I think it's a useful subject to dig into and get more opinions on.
https://github.com/TestAnything/Specification/issues/41
